---
title: "Untitled"
author: "Gaurav"
date: "April 8, 2019"
output: html_document
---

#Generate Reads
```{r}
generateReads <- function(genome,depth = 30, readLen = 5000 ){
set.seed(123)
genomeLen <- nchar(genome)
nreads <- genomeLen*depth/readLen
startPos <- sample(genomeLen,nreads)
reads <- NULL
for(i in 1:nreads){
      if(startPos[i] + (readLen - 1) > genomeLen){
            reads <- c(reads,substr(genome,startPos[i],genomeLen))
      }
      else{
            reads <- c(reads,substr(genome,startPos[i],(startPos[i] + readLen - 1)))
      }
      
}
return(reads)
}
```

```{r}
error = 0.1
nucleotides <- c('A','T','C','G')
```
flipReads
```{r flipReads}
flipReads <- function(rds,error){
flippedReads <- NULL
for(i in seq(1,length(rds))){
subread <- rds[i]
subreadLen <- nchar(subread)
nErrors <- floor(subreadLen * error)
subread <- strsplit(subread,'')[[1]]
if (nErrors != 0) { subread <- flipBase(subread,subreadLen,nErrors)}
flippedReads <- c(flippedReads, paste0(subread,collapse = ''))
}
return(flippedReads)
}
```

```{r}
flipBase <- function(read,readLen,nErr){
samp <- sample(readLen,nErr)
for(i in 1:length(samp)){
      basePos <- which(nucleotides == read[samp[i]])
      read[samp[i]] <- sample(nucleotides[-basePos],1)
}
return(read)
}
```

```{r}
for(i in 5){
print(i)
flippedReads <- reads
flippedReads <- flipReads(flippedReads,i/100)
print(sum(strsplit(reads[2],'')[[1]] != strsplit(flippedReads[2],'')[[1]]))
readID <- paste0('@read_',1:nreads,'\n')
readIdNflpdReads <- paste0(readID,toupper(flippedReads))
setwd("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try")
filename = paste0("flippedReadsWithError",as.character(i),"Pcent1L.fq")
write(readIdNflpdReads,file = filename)
}
#write(readIdNflpdReads5,file = 'ecoli_5pcent_error_reads.fasta')
```

```{r}
#Count kmers
# kmerLen = 21
# kmerEndPos = nchar(ecoliGenome) - kmerLen + 1
# kmers <- NULL
# for(i in 1:kmerEndPos){
#       kmers <- c(kmers,substr(ecoliGenome,i,i+20))
# }
# kmerFreq <- table(kmers)
# beepr::beep(8)
```
getKmersFromRead
```{r getKmersFromRead}
#kmers for reads
getKmersFromRead <- function(flippedReads,kmerLen){
readKmers <- NULL
for(i in 1:length(flippedReads)){
      if(nchar(flippedReads[i]) > kmerLen){
      kmerEndPos = nchar(flippedReads[i]) - kmerLen + 1
      for(j in 1:kmerEndPos){
            readKmers <- c(readKmers,substr(flippedReads[i],j,j+kmerLen-1))
      }}
#freqReadKmers <- table(readKmers)
}
beepr::beep(8)
return(readKmers)
}
```
Read Ecoli Genome
```{r}
library(seqinr)
read.fasta('D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/ecoliGenome.fasta') -> ecoliSeq
# ecoliSeq
# ecoliSeq$NC_011750.1
View(ecoliSeq)
paste0(ecoliSeq$NC_011750.1)
paste0(ecoliSeq$NC_011750.1,collapse = '') -> ecoliGenome
#nchars(ecoliGenome)
length(ecoliGenome)
ecoliGenome <- ecoliGenome[1]
length(ecoliGenome)
#nchars(ecoliGenome)
nchar(ecoliGenome)
ecoli1L <- substr(ecoliGenome,1,1e5)
```

```{r}
setwd("D:/JHU/Spring19/601.749ComparativeGenomics/Project")
kmerDf <- list()
num <- 1
dfKmers <- data.frame(count = 1:210)
for(i in seq(0,15,3)){
      filename <- paste0(as.character(i),'pcent21mer.histo')
      kmerDf[[num]] <- read.table(filename,sep = ' ',header = F)
      dfKmers[[paste0('err',as.character(i))]] <- 0
      dfKmers[,(num + 1)][which(dfKmers$count %in% kmerDf[[num]]$V1)] <- kmerDf[[num]]$V2
      num <- num + 1
}
```

```{r}
library(reshape2)
library(ggplot2)
dfKmersMelt <- melt(dfKmers, id.vars = 1)
colnames(dfKmersMelt)[2] <- "Error"
ggplot(dfKmersMelt,aes(x=count,y=log10(value),colour = Error)) + geom_line() + xlab("Kmer Count") + ylab("Log10(#kmers)") + labs(color = "Error %\n")
```
#generate kmers from the 1st 1000 bases
```{r}
ecoli1k <- substr(ecoli1M,1,100000)
kmers <- NULL
kmerEndPos = nchar(ecoli1k) - 20
for(i in seq(1,kmerEndPos,20)){
      kmers <- c(kmers, substr(ecoli1k,i,i+20))
}
```

```{r}
library(stringdist)
kmerDist <- stringdistmatrix(kmers)
kmerMat <- as.matrix(kmerDist)
table(kmerMat)[1:5]
```
Found only adjoining kmers have smaller distance of 2 with each other 

Looking at 1000 mers
```{r}
reads1000mers <- getKmersFromRead(reads[1:10],1000)
```

```{r}
read1000mersDist <- stringdistmatrix(reads1000mers)
```

```{r message=F}
filepath <- "D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/4pcent21mer1L.fa"
readKmerfile <- function(filepath)
{
pcent0_21mer <- read.table(filepath,stringsAsFactors = F )
library(dplyr)
crt <- grepl('>',pcent0_21mer[,1])
p0m21_1 <- pcent0_21mer[crt,]
freq_p021m <- sapply(seq(1,length(p0m21_1)), function(x){strsplit(p0m21_1[x],'>',fixed = T)[[1]][2]}) %>% as.numeric()
p0m21_2 <- pcent0_21mer[!crt,]
p0m21 <- data.frame(kmer = p0m21_2, freq = freq_p021m, stringsAsFactors = F)
return(p0m21)
}
#p0m21 <- as.data.frame(cbind(p0m21_1,p0m21_2))
# rm(p0m21_2)
# rm(p0m21_1)
# rm(freq_p021m)
# rm(crt)
# rm(pcent0_21mer)
```

```{r}
library(hash)
#options(expressions = 1e6)
kmerHash <- hash(p0m21[,1],p0m21[,2])
```
```{r}
write.table(p0m21,file='p0m21.csv',quote = F, row.names = F, sep = ',')
```
Mutate and compare
```{r}
kmerLen <- 21
sameKmer <- NULL
n <- length(p0m21[,1])
nucleotidesCap <- list(A = 'A',C = 'C',G = 'G',T ='T')
for(i in seq(1,10000)){
      splitKmer <- strsplit(as.character(p0m21[i,1]),'',fixed = T)[[1]]
      tempSameKmer <- 0
      for(j in seq(1,kmerLen)){
            #print(tempSameKmer)
            nucList <- nucleotidesCap
            nucList[[splitKmer[j]]] <- NULL
            if(j == 1){
                  merged <- paste0(splitKmer[2:kmerLen],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(nucList[x],merged,collapse = '')})
            }
            else if(j == kmerLen){
                  merged <- paste0(splitKmer[1:(kmerLen-1)],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(merged,nucList[x],collapse = '')})
            }
            else {
                  merged1 <- paste0(splitKmer[1:(j-1)],collapse = '')
                  merged2 <- paste0(splitKmer[(j+1):kmerLen],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(merged1,nucList[x],merged2,collapse = '')})
            }
            tempSameKmer <- tempSameKmer + sum(mutatedKmers %in% p0m21[,1])
      }
      sameKmer <- c(sameKmer,tempSameKmer)
      frac <- i/n
      if(frac > 0.95){
            print("95% done")
      }
      else if(frac > 0.5){
            print("50% done")
      }
      else if(frac > 0.1){
            print("10% done")
      }
            
}
```

```{r}
splitAt <- function(x, pos) unname(split(x, cumsum(seq_along(x) %in% pos)))
```

```{r}
library(stringdist)
distMat <- stringdistmatrix(as.character(p5m21[1:100000,1]),method = "hamming")
```

```{r}
p5m21 <- readKmerfile("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/4pcent21mer1L.fa")
```

```{r}
indices <- list()
num <- 1
for(i in seq(1,900000,10000)){
distMat <- stringdistmatrix(as.character(p5m21[i:(i+9999),1]),method = "hamming")
indices[[num]] <- which(as.matrix(distMat) == 1)
num <- num + 1
if(num > 90){
      print("90% done")
}
else if(num > 50){
      print("50% done")
}
else if(num > 10){
      print("10% done")
}
}
```

```{r}
rows <- NULL
cols <- NULL
num <- 0
for(j in 1:length(indices)){
indx <- indices[[j]]
for(i in seq(1:length(indx))){
rowNo <- (indx[i] %% 10000) + num*10000
colNo <- ceiling(indx[i]/10000) + num*10000
if(rowNo >= colNo){
      rows <- c(rows,rowNo)
      cols <- c(cols,colNo)
}
}
num <- num +1
}
```
Rows and cols are indices of kmers with 1 hamming distance
```{r}
df <- data.frame(row = rows, col = cols, stringsAsFactors = F)
```

Note the dimension of simKmers will look more because df is getting trimmed
```{r}
simKmers <- list()
num <- 1
for(i in 1:length(df[,1])){
      kmerIndices <- NULL
      howManyKmersEqual <- which(df[,1] == df[i,1])
      if(length(howManyKmersEqual) > 1){
            simKmers[[num]] <- df[,2][howManyKmersEqual[-1]]
            df <- df[-howManyKmersEqual[-1],]
      }
      else {
            simKmers[[num]] <- 0
      }
      num <- num + 1
}
```


```{r}
nonZeroSimKmers <- sapply(1:length(simKmers),function(x) simKmers[[x]][1] != 0)
```

Now add kmer freq
```{r}
dfm <- as.matrix(df)
dfm <- dfm[,c(2,1)]
for(i in 1:length(simKmers)){
      if(simKmers[[i]][1] != 0){
      simKmersIndices <-  as.vector(c(dfm[i,],simKmers[[i]][1]))
      }
      else {
            simKmersIndices <- as.vector(dfm[i,])
      }
      p5m21$freq[simKmersIndices[1]] <- sum(p5m21$freq[simKmersIndices])
      p5m21 <- p5m21[-simKmersIndices[-1],]
}
```

```{r}
table(p5m21$freq) -> tb
write.csv(as.data.frame(tb),file = '4p21m1Liter1.csv',row.names = F,quote = F)
```
Ecoli 10 k reads of length 1k
```{r}
ecoli10k <- substr(ecoli1L,1,1e4)
reads10k <- generateReads( genome = ecoli10k, readLen = 1000)
#flipReads - Generate Errors
reads10kWith5PcentError <- flipReads(reads10k,0.05)
reads10kWith5PcentError <- toupper(reads10kWith5PcentError)
reads10kWith10PcentError <- flipReads(reads10k,0.1)
reads10kWith15PcentError <- flipReads(reads10k,0.15)
reads10kWith20PcentError <- flipReads(reads10k,0.2)
```

```{r}
for(i in 5){
print(i)
nreads <- length(reads10k)
#nreads <- 300
flippedReads <- reads10k
flippedReads <- flipReads(flippedReads,i/100)
print(sum(strsplit(reads10k[2],'')[[1]] != strsplit(flippedReads[2],'')[[1]]))
readID <- paste0('>read_',1:nreads)
readIdNflpdReads <- unlist(lapply(1:nreads,function(x){c(readID[x],toupper(flippedReads[x]))}))
writeDf <- data.frame(data = readIdNflpdReads)
setwd("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/10k")
filename = paste0("flippedReadsWithError",as.character(i),"Pcent10k.fasta")
write.table(readIdNflpdReads,file = filename, col.names = F, row.names = F, quote = F)
}
```
Get kmers using parallel
```{r}
getKmersFromRead <- function(flippedReads,kmerLen = 21){
readLen <- nchar(flippedReads)
readKmers <- sapply(1:(readLen - kmerLen +1),function(i){
            substr(flippedReads,i,i+kmerLen-1)
      })
#freqReadKmers <- table(readKmers)
#beepr::beep(8)
return(readKmers)
}
```
getKmers
```{r}
kmers10kWith5pcentError <- sapply(1:length(reads10k),function(x) getKmersFromRead(reads10kWith5PcentError[x]))
kmers10k <- sapply(1:length(reads10k),function(x) getKmersFromRead(reads10k[x]))
kmers10k <- unlist(kmers10k)
beepr::beep(8)
kmers10kWith5pcentError <- unlist(kmers10kWith5pcentError) #Unlist kmers
#get Table
kmers10kE05X30T <- table(kmers10kWith5pcentError) 
#get Df
kmers10kE05X30Df <- as.data.frame(kmers10kE05X30T, stringAsFactors = FALSE )
#remove Table
rm(kmers10kE05X30T)
#for some reasons the table was giving 4 kmers extra
```

```{r getKmersDf}
getKmersDf <- function(reads10kWith5PcentError){
nReads <- length(reads10kWith5PcentError)
kmers10kWith5pcentError <- sapply(1:nReads,function(x) getKmersFromRead(reads10kWith5PcentError[x]))
kmers10kWith5pcentError <- unlist(kmers10kWith5pcentError) #Unlist kmers
#get Table
kmers10kE05X30T <- table(kmers10kWith5pcentError) 
#get Df
kmers10kE05X30Df <- as.data.frame(kmers10kE05X30T, stringAsFactors = FALSE )
#remove Table
rm(kmers10kE05X30T)
beepr::beep(8)
return(kmers10kE05X30Df)
}
```

```{r}
kmers10kE00X30Df <- getKmersDf(reads10k)
kmers10kWith10PcentErrorDf <- getKmersDf(reads10kWith10PcentError)
kmers10kWith15PcentErrorDf <- getKmersDf(reads10kWith15PcentError)
kmers10kWith20PcentErrorDf <- getKmersDf(reads10kWith20PcentError)
```
getHammingDistances from reference string
```{r}
library(stringdist)
refKmer <- as.character(kmers10kE05X30Df[1,1])
refKmer <- paste0(rep('A',21),collapse = '')
#refKmer <- paste0(refKmer,'A',collapse = '')
# refKmer <- 'AAAAACTGCTGGCTGTGGGGC'
refDist2 <- sapply(1:nrow(kmers10kE05X30Df),function(x){stringdist(kmers10kE05X30Df[x,1],refKmer,method = "hamming")})
table(refDist2)
```
get distances from ref string
```{r}
refKmer <- paste0(rep('A',21),collapse = '')
refDist00p30X <- sapply(1:nrow(kmers10kE00X30Df),function(x){stringdist(kmers10kE00X30Df[x,1],refKmer,method = "hamming")})
refDist10kE05X30 <- sapply(1:nrow(kmers10kE05X30Df),function(x){stringdist(kmers10kE05X30Df[x,1],refKmer,method = "hamming")})
refDist10PcentError <- sapply(1:nrow(kmers10kWith10PcentErrorDf),function(x){stringdist(kmers10kWith10PcentErrorDf[x,1],refKmer,method = "hamming")})
refDist15PcentError <- sapply(1:nrow(kmers10kWith15PcentErrorDf),function(x){stringdist(kmers10kWith15PcentErrorDf[x,1],refKmer,method = "hamming")})
refDist20PcentError <- sapply(1:nrow(kmers10kWith20PcentErrorDf),function(x){stringdist(kmers10kWith20PcentErrorDf[x,1],refKmer,method = "hamming")})
beepr::beep(8)
```
Make Tables
```{r}
refDist10kT <- table(refDist00p30X)
refDist5pT <- table(refDist)
refDist10pT <- table(refDist10PcentError)
refDist15pT <- table(refDist15PcentError)
refDist20pT <- table(refDist20PcentError)
```

```{r}
library(ggplot2)
addFreqCol <- function(df,refDistTable,colname){
   df[[colname]] <- 0
   tableIdx <- names(refDistTable) %in% df$dist
   dfIdx <- which(df$dist %in% names(refDistTable))
   df[dfIdx,colname] <- refDistTable[tableIdx]
   return(df)
}
df <- data.frame(dist = 5:21)
df <- addFreqCol(df,refDist10kT,'0')
df <- addFreqCol(df,refDist5pT,'5')
df <- addFreqCol(df,refDist10pT,'10')
df <- addFreqCol(df,refDist15pT,'15')
df <- addFreqCol(df,refDist20pT,'20')
library(reshape2)
dfm <- melt(df,id.vars = 1,measure.vars = 2:6)
```
```{r}
ggplot(dfm,aes(x = dist, y = value, colour = variable)) + geom_line() + labs(colour = "Percentage Error") + theme(plot.title = element_text(hjust = 0.5))
```

cluster similar kmers together
```{r}
ref16Df <- kmers10kE05X30Df[which(refDist == 16),]
ref16Kmer <- ref16Df[1,1]
ref16Dist <- sapply(1:nrow(ref16Df),function(x){stringdist(ref16Df[x,1],ref16Kmer,method = "hamming")})
table(ref16Dist)
```

```{r}
library(qrqc)
reads10kFile5pcent <- readSeqFile("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/10k/flippedReadsWithError0Pcent10k.fa",type = "fasta",max.length = 5000, k = 21, quality = "sanger")
```
Try the same with 50x depth
```{r}
reads10kE00X50 <- generateReads(genome = ecoli10k, readLen = 1000, depth = 50)
#flipReads - Generate Errors
reads10kE05X50 <- flipReads(reads10kE00X50,0.05)
reads10kE10X50 <- flipReads(reads10kE00X50,0.1)
reads10kE15X50 <- flipReads(reads10kE00X50,0.15)
reads10kE20X50 <- flipReads(reads10kE00X50,0.2)
```
getKmers
```{r}
kmers10kE00X50Df <- getKmersDf(reads10kE00X50)
kmers10kE05X50Df <- getKmersDf(reads10kE05X50)
kmers10kE10X50Df <- getKmersDf(reads10kE10X50)
kmers10kE15X50Df <- getKmersDf(reads10kE15X50)
kmers10kE20X50Df <- getKmersDf(reads10kE20X50)
```
getHammingDistances from reference string
```{r}
library(stringdist)
refKmer <- paste0(rep('A',21),collapse = '')
```
get distances from ref string
```{r}
refKmer <- paste0(rep('A',21),collapse = '')
refDist10kE00X50 <- sapply(1:nrow(kmers10kE00X50Df),function(x){stringdist(kmers10kE00X50Df[x,1],refKmer,method = "hamming")})
refDist10kE05X50 <- sapply(1:nrow(kmers10kE05X50Df),function(x){stringdist(kmers10kE05X50Df[x,1],refKmer,method = "hamming")})
refDist10kE10X50 <- sapply(1:nrow(kmers10kE10X50Df),function(x){stringdist(kmers10kE10X50Df[x,1],refKmer,method = "hamming")})
refDist10kE15X50 <- sapply(1:nrow(kmers10kE15X50Df),function(x){stringdist(kmers10kE15X50Df[x,1],refKmer,method = "hamming")})
refDist10kE20X50 <- sapply(1:nrow(kmers10kE20X50Df),function(x){stringdist(kmers10kE20X50Df[x,1],refKmer,method = "hamming")})
beepr::beep(8)
```
Make Tables
```{r}
refDist00k50XT <- table(refDist10kE00X50)
refDist05p50XT <- table(refDist10kE05X50)
refDist10p50XT <- table(refDist10kE10X50)
refDist15p50XT <- table(refDist10kE15X50)
refDist20p50XT <- table(refDist10kE20X50)
```

```{r}
df50X <- data.frame(dist = 5:21)
df50X <- addFreqCol(df50X,refDist00k50XT,'0')
df50X <- addFreqCol(df50X,refDist05p50XT,'5')
df50X <- addFreqCol(df50X,refDist10p50XT,'10')
df50X <- addFreqCol(df50X,refDist15p50XT,'15')
df50X <- addFreqCol(df50X,refDist20p50XT,'20')
library(reshape2)
dfm50X <- melt(df50X,id.vars = 1,measure.vars = 2:6)
```
```{r}
ggplot(dfm50X,aes(x = dist, y = value, colour = variable)) + geom_line() + labs(colour = "Percentage Error") + theme(plot.title = element_text(hjust = 0.5))
```
check 0% error spectrum genomescope 100bp reads
Even though Genomescope fails to converge it estimates the coverage correctly but only with zero error

Now coming back to find the same, count the name
```{r}
kmers10kE05X30Df$refDist <- refDist10kE05X30
kmers10kE05X30Df <- kmers10kE05X30Df[-which(kmers10kE05X30Df$refDist == Inf),]
```
Add index column and checking same distance matrix
```{r}
library(dplyr)
kmers10kE05X30Df <- kmers10kE05X30Df %>% arrange(refDist)
kmers10kE05X30Df$index <- seq(1,nrow(kmers10kE05X30Df))
kmers10kE05X30T <- table(kmers10kE05X30Df$refDist)
km10kE05X30D67 <- kmers10kE05X30Df %>% filter(refDist == 6 | refDist == 7)
dim(km10kE05X30D67)
```
```{r}
distMatKm10kE05X30D67 <- stringdistmatrix(km10kE05X30D67$kmers10kWith5pcentError, method = "hamming")
```

```{r}
dist1 <- which(as.matrix(distMatKm10kE05X30D67) == 1)
#later add the previous rows but for here since it is the first 2 clusters don't need to add
dimMat <- nrow(as.matrix(distMatKm10kE05X30D67))
rowNo <- dist1 %% dimMat
colNo <- ceiling(dist1/dimMat)
rc <- data.frame(r = rowNo, c = colNo)
rc <- rc[rc$r > rc$c,]
```
#Test out scoring scheme ATGC = 1,2,4,5
```{r}
scoring = list(A = 1, T = 2, C = 4, G = 5)
getScore <- function(seq){
seqBases <- strsplit(seq,'')[[1]]
score <- sapply(1:length(seqBases),function(x) {
   scoring[[seqBases[x]]]
   })
return(sum(score))
}
#Add connected column
kmers10kE05X30Df$connected <- rep('0',nrow(kmers10kE05X30Df))
```
Add connected components
```{r}
sapply(1:nrow(rc),function(x){
   #print(x)
   if(kmers10kE05X30Df$connected[rc$c[x]] == '0'){
      kmers10kE05X30Df$connected[rc$c[x]] <<- as.character(rc$r[x])
      #print(kmers10kE05X30Df$connected[rc$c[x]])
   }
   else {
      kmers10kE05X30Df$connected[rc$c[x]] <<- paste0(kmers10kE05X30Df$connected[rc$c[x]],',',as.character(rc$r[x]))
      #print(kmers10kE05X30Df$connected[rc$c[x]])
   }
}
   )
```

## Try Clustering on kmers
AAA
AAT
GGG
GGA
AAG
```{r}
bases <- list(A = c(1,0,0,0),C = c(0,1,0,0),G = c(0,0,1,0),T = c(0,0,0,1))
```

```{r getSeqVector}
getSeqVector <- function(seq){
seqBases <- strsplit(seq,'')[[1]]
seqVector <- NULL
sv <- sapply(1:length(seqBases),function(x) {
   seqVector <<- c(seqVector,bases[[seqBases[x]]])
   })
return(seqVector)
}
```

```{r}
seqs <- c('AAA','AAT','GGG','GGA','AAG','CCA','GCC','AAC','CCC','TTT','GAC','TAT','CGG')
seqMatrix <- sapply(1:length(seqs), function(x) getSeqVector(seqs[x]))
seqMatrix <- t(seqMatrix)
```

```{r}
library(umap)
umSM <- umap(seqMatrix, method = "umap-learn")
```

```{r}
udf <- as.data.frame(umSM$layout)
udf$seqs <- seqs
udf$index <- as.factor(1:13)
ggplot(udf,aes(x = V1, y = V2, colour = seqs )) + geom_point() + geom_text(aes(label = seqs)) + labs(title = "UMAP of 3mers")
```

