---
title: "Untitled"
author: "Gaurav"
date: "April 8, 2019"
output: html_document
---

```{r}
set.seed(123)
startPos <- sample(1e6,nreads)
reads <- NULL
for(i in 1:nreads){
      if(startPos[i] + 9999 > 1e6){
            reads <- c(reads,substr(ecoli1M,startPos[i],1e6))
      }
      else{
            reads <- c(reads,substr(ecoli1M,startPos[i],startPos[i] + 9999))
      }
      
}
```

```{r}
error = 0.1
nucleotides <- c('a','t','c','g')
```

```{r}
flipReads <- function(rds,error){
flippedReads <- NULL
for(i in seq(1,length(rds))){
subread <- rds[i]
subreadLen <- nchar(subread)
nErrors <- floor(subreadLen * error)
subread <- strsplit(subread,'')[[1]]
subread <- flipBase(subread,subreadLen,nErrors)
flippedReads <- c(flippedReads, paste0(subread,collapse = ''))
}
return(flippedReads)
}
```

```{r}
flipBase <- function(read,readLen,nErr){
samp <- sample(readLen,nErr)
for(i in 1:length(samp)){
      basePos <- which(nucleotides == read[samp[i]])
      read[samp[i]] <- sample(nucleotides[-basePos],1)
}
return(read)
}
```

```{r}
for(i in 6:20){
print(i)
flippedReads <- reads
flippedReads <- flipReads(flippedReads,i/100)
print(sum(strsplit(reads[2],'')[[1]] != strsplit(flippedReads[2],'')[[1]]))
readID <- paste0('@read_',1:3000,'\n')
readIdNflpdReads <- paste0(readID,toupper(flippedReads))
setwd("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try")
filename = paste0("flippedReadsWithError",as.character(i),"Pcent.fq")
write(readIdNflpdReads,file = filename)
}
#write(readIdNflpdReads5,file = 'ecoli_5pcent_error_reads.fasta')
```
```{r}
#Count kmers
kmerLen = 21
kmerEndPos = nchar(ecoliGenome) - kmerLen + 1
kmers <- NULL
for(i in 1:kmerEndPos){
      kmers <- c(kmers,substr(ecoliGenome,i,i+20))
}
kmerFreq <- table(kmers)
beepr::beep(8)
```

```{r}
#kmers for reads
getKmersFromRead <- function(flippedReads,kmerLen){
readKmers <- NULL
for(i in 1:length(flippedReads)){
      if(nchar(flippedReads[i]) > kmerLen){
      kmerEndPos = nchar(flippedReads[i]) - kmerLen + 1
      for(j in 1:kmerEndPos){
            readKmers <- c(readKmers,substr(flippedReads[i],j,j+kmerLen-1))
      }}
#freqReadKmers <- table(readKmers)
}
beepr::beep(8)
return(readKmers)
}
```


```{r}
read.fasta('data/ecoliGenome.fasta') -> ecoliSeq
ecoliSeq
ecoliSeq$NC_011750.1
View(ecoliSeq)
paste0(ecoliSeq$NC_011750.1)
paste0(ecoliSeq$NC_011750.1,collapse = '') -> ecoliGenome
nchars(ecoliGenome)
length(ecoliGenome)
ecoliGenome <- ecoliGenome[1]
length(ecoliGenome)
nchars(ecoliGenome)
nchar(ecoliGenome)
ecoli1M <- substr(ecoliGenome,1,1e6)
```

```{r}
setwd("D:/JHU/Spring19/601.749ComparativeGenomics/Project")
kmerDf <- list()
num <- 1
dfKmers <- data.frame(count = 1:210)
for(i in seq(0,15,3)){
      filename <- paste0(as.character(i),'pcent21mer.histo')
      kmerDf[[num]] <- read.table(filename,sep = ' ',header = F)
      dfKmers[[paste0('err',as.character(i))]] <- 0
      dfKmers[,(num + 1)][which(dfKmers$count %in% kmerDf[[num]]$V1)] <- kmerDf[[num]]$V2
      num <- num + 1
}
```
```{r}
library(reshape2)
library(ggplot2)
dfKmersMelt <- melt(dfKmers, id.vars = 1)
colnames(dfKmersMelt)[2] <- "Error"
ggplot(dfKmersMelt,aes(x=count,y=log10(value),colour = Error)) + geom_line() + xlab("Kmer Count") + ylab("Log10(#kmers)") + labs(color = "Error %\n")
```
#generate kmers from the 1st 1000 bases
```{r}
ecoli1k <- substr(ecoli1M,1,100000)
kmers <- NULL
kmerEndPos = nchar(ecoli1k) - 20
for(i in seq(1,kmerEndPos,20)){
      kmers <- c(kmers, substr(ecoli1k,i,i+20))
}
```

```{r}
library(stringdist)
kmerDist <- stringdistmatrix(kmers)
kmerMat <- as.matrix(kmerDist)
table(kmerMat)[1:5]
```
Found only adjoining kmers have smaller distance of 2 with each other 

Looking at 1000 mers
```{r}
reads1000mers <- getKmersFromRead(reads[1:10],1000)
```

```{r}
read1000mersDist <- stringdistmatrix(reads1000mers)
```

```{r message=F}
filepath <- "D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/0pcent21mer.fa"
readKmerfile <- function(filepath)
{
pcent0_21mer <- read.table(filepath,stringsAsFactors = F )
library(dplyr)
crt <- grepl('>',pcent0_21mer[,1])
p0m21_1 <- pcent0_21mer[crt,]
freq_p021m <- sapply(seq(1,length(p0m21_1)), function(x){strsplit(p0m21_1[x],'>',fixed = T)[[1]][2]}) %>% as.numeric()
p0m21_2 <- pcent0_21mer[!crt,]
p0m21 <- data.frame(kmer = p0m21_2, freq = freq_p021m, stringsAsFactors = F)
return(p0m21)
}
#p0m21 <- as.data.frame(cbind(p0m21_1,p0m21_2))
# rm(p0m21_2)
# rm(p0m21_1)
# rm(freq_p021m)
# rm(crt)
# rm(pcent0_21mer)
```

```{r}
library(hash)
#options(expressions = 1e6)
kmerHash <- hash(p0m21[,1],p0m21[,2])
```
```{r}
write.table(p0m21,file='p0m21.csv',quote = F, row.names = F, sep = ',')
```
Mutate and compare
```{r}
kmerLen <- 21
sameKmer <- NULL
n <- length(p0m21[,1])
nucleotidesCap <- list(A = 'A',C = 'C',G = 'G',T ='T')
for(i in seq(1,10000)){
      splitKmer <- strsplit(as.character(p0m21[i,1]),'',fixed = T)[[1]]
      tempSameKmer <- 0
      for(j in seq(1,kmerLen)){
            #print(tempSameKmer)
            nucList <- nucleotidesCap
            nucList[[splitKmer[j]]] <- NULL
            if(j == 1){
                  merged <- paste0(splitKmer[2:kmerLen],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(nucList[x],merged,collapse = '')})
            }
            else if(j == kmerLen){
                  merged <- paste0(splitKmer[1:(kmerLen-1)],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(merged,nucList[x],collapse = '')})
            }
            else {
                  merged1 <- paste0(splitKmer[1:(j-1)],collapse = '')
                  merged2 <- paste0(splitKmer[(j+1):kmerLen],collapse = '')
                  mutatedKmers <- sapply(1:3,function(x){paste0(merged1,nucList[x],merged2,collapse = '')})
            }
            tempSameKmer <- tempSameKmer + sum(mutatedKmers %in% p0m21[,1])
      }
      sameKmer <- c(sameKmer,tempSameKmer)
      frac <- i/n
      if(frac > 0.95){
            print("95% done")
      }
      else if(frac > 0.5){
            print("50% done")
      }
      else if(frac > 0.1){
            print("10% done")
      }
            
}
```

```{r}
splitAt <- function(x, pos) unname(split(x, cumsum(seq_along(x) %in% pos)))
```

```{r}
library(stringdist)
distMat <- stringdistmatrix(as.character(p0m21[1:10000,1]),method = "hamming")
```

```{r}
p5m21 <- readKmerfile("D:/JHU/Spring19/601.749ComparativeGenomics/Project/bioSeqGen/data/try/5pcent21mer.fa")
```

```{r}
indices <- list()
num <- 1
for(i in seq(1,1060000,10000)){
distMat <- stringdistmatrix(as.character(p5m21[i:(i+9999),1]),method = "hamming")
indices[[num]] <- which(as.matrix(distMat) == 1)
num <- num + 1
if(num > 90){
      print("90% done")
}
else if(num > 50){
      print("50% done")
}
else if(num > 10){
      print("10% done")
}
}
```

```{r}
rows <- NULL
cols <- NULL
num <- 0
for(j in 1:2){
indx <- indices[[j]]
for(i in seq(1:length(indx))){
rowNo <- (indx[i] %% 10000) + num*10000
colNo <- ceiling(indx[i]/10000) + num*10000
if(rowNo >= colNo){
      rows <- c(rows,rowNo)
      cols <- c(cols,colNo)
}
}
num <- num +1
}
```

```{r}
p5m21copy <- p5m21
```



